Add energy and related quantities.

Aehmlo opened the pull request:
This closes #66, #67, #69, and #74 by adding the respective quantities and units. It also closes #75 and fixes some incorrect labels in `electric_current.rs`.

iliekturtles commented on src/si/available_energy.rs:
```diff
@@ -0,0 +1,25 @@
+quantity! {
+    /// Available energy (base unit joule per kilogram, m<sup>2</sup> · s<sup>-2</sup>).
+    quantity: AvailableEnergy; "available energy";
+    /// Available energy dimension, m<sup>2</sup> · s<sup>-2</sup>).
```
Extra `)` at the end of the line.

iliekturtles commented on src/si/electric_potential.rs:
```diff
@@ -0,0 +1,42 @@
+//! Electric potential (base unit volt, m<sup>2</sup> · kg · s<sup>-3</sup> · V<sup>-1</sup>).
```
Last dimension should be `A<sup>-1</sup>`.

iliekturtles commented on src/si/energy.rs:
```diff
@@ -0,0 +1,63 @@
+quantity! {
+    /// Energy (base unit joule, m<sup>2</sup> · kg · s<sup>-2</sup>).
+    quantity: Energy; "energy";
+    /// Energy dimension, m<sup>2</sup> · kg · s<sup>-2</sup>).
```
Extra `)` at the end.

iliekturtles commented:
Build failures are timeouts. Pretty sure this is #52. I don't want to hold up this PR based on #52 so some changes to the CI scripts may be necessary.

iliekturtles commented:
- Energy
  - "Derived unit of energy" doc comment should be on `joule` instead of `microjoule`.
  - `microjoule` through `yoctojoule` need their abbreviations and descriptions corrected.
  - Replace all "masculine ordinal indicator" (º) with "degree sign" (°).
  - Add an `_` before the `E` in the conversion factor for `foot_pound`, `kilocalorie`, `kilowatt_hour`, and `therm_ec`.
  - Change Btu descriptions to be spelled out: `"British thermal unit ...", "British thermal units ..."`.
  - Change `calorie_15` descriptions to include `"(15 °C)"`. Similar for `calorie_20`.
  - Adjust `foot_poundal` abbreviation to be `"ft · pdl"`.
  - Adjust `foot_poundal` descriptions to be `"foot poundal", "foot poundals"`.
  - Adjust `foot_pound` abbreviation to be `"ft · lbf"`.
  - Adjust `foot_pound` descriptions to be `"foot pound-force", "foot pounds-force"`.
  - Adjust `kilowatt_hour abbreviation to be `"kW · h"`.
  - Adjust 'kilowatt_hour descriptions to be `"kilowatt hour", "kilowatt hours"`.
  - Adjust `ton_tnt` abbreviation to `"t of TNT"`.
  - Adjust `watt_hour` abbreviation to be `"W · h"`.
  - Adjust `watt_hour` descriptions to be `"watt hour", "watt hours"`.
  - Add `watt_second` with a conversion factor of `1.0_E0`.

Aehmlo commented:
Energy was a haul; I'm not surprised to find these errors, but I think I've fixed them all now.

Currently working to write some tests; I'll add a commit to this branch when that's done.

Aehmlo commented:
I've added tests for electric potential, energy, and available energy.

iliekturtles commented:
Energy took a while to review. I appreciate all the effort you've put in to put it together.

- Energy
  - See diff below. Wrapped long lines. Added missing Celsius abbreviations. Corrected J * s to be W * s in unit name and descriptions. Simplified test formulas.
- Available energy
  - Spell out `"British thermal unit"` in descriptions.
  - Change `btu_it_per_pound` abbreviation to `Btu (IT)/lb`.
  - Change `calories_it_per_pound` to `calorie_it_per_pound`
  - Change `calories_it_per_pound` abbreviation to `cal (IT)/lb`.
  - Change `calories_it_per_pound` descriptions to say "pound" instead of "gram".
  - Change `calories_per_pound` to `calorie_per_pound`
  - Change `calories_per_pound` abbreviation to `cal/lb`.
  - Change `calories_per_pound` descriptions to say "pound" instead of "gram".
  - Change test formulas similar to changes in the diff for energy.

Energy:
```diff
diff --git a/src/si/energy.rs b/src/si/energy.rs
index 1f02633..1089e6f 100644
--- a/src/si/energy.rs
+++ b/src/si/energy.rs
@@ -36,30 +36,35 @@ quantity! {
         @zeptojoule: prefix!(zepto); "zJ", "zeptojoule", "zeptojoules";
         @yoctojoule: prefix!(yocto); "yJ", "yoctojoule", "yoctojoules";
 
-        @btu_it: 1.055_056_E3; "Btu (IT)", "British thermal unit (IT)", "British thermal units (IT)";
+        @btu_it: 1.055_056_E3; "Btu (IT)", "British thermal unit (IT)",
+            "British thermal units (IT)";
         @btu: 1.054_350_E3; "Btu", "British thermal unit", "British thermal units";
-        @btu_39: 1.059_67_E3; "Btu₃₉", "British thermal unit (39 °F)", "British thermal units (39 °F)";
-        @btu_59: 1.054_80_E3; "Btu₅₉", "British thermal unit (59 °F)", "British thermal units (59 °F)";
-        @btu_60: 1.054_68_E3; "Btu₆₀", "British thermal unit (60 °F)", "British thermal units (60 °F)";
+        @btu_39: 1.059_67_E3; "Btu₃₉", "British thermal unit (39 °F)",
+            "British thermal units (39 °F)";
+        @btu_59: 1.054_80_E3; "Btu₅₉", "British thermal unit (59 °F)",
+            "British thermal units (59 °F)";
+        @btu_60: 1.054_68_E3; "Btu₆₀", "British thermal unit (60 °F)",
+            "British thermal units (60 °F)";
         @calorie_it: 4.186_8_E0; "cal (IT)", "calorie (IT)", "calories (IT)";
         @calorie: 4.184_E0; "cal", "calorie", "calories";
-        @calorie_15: 4.185_80_E0; "cal₁₅", "calorie (15°)", "calories (15 °C)";
-        @calorie_20: 4.181_90_E0; "cal₂₀", "calorie (20°)", "calories (20 °C)";
+        @calorie_15: 4.185_80_E0; "cal₁₅", "calorie (15 °C)", "calories (15 °C)";
+        @calorie_20: 4.181_90_E0; "cal₂₀", "calorie (20 °C)", "calories (20 °C)";
         @calorie_it_nutrition: 4.186_8_E3; "Cal (IT)", "Calorie (IT)", "Calories (IT)";
         @calorie_nutrition: 4.184_E3; "Cal", "Calorie", "Calories";
         @electronvolt: 1.602_177_E-19; "eV", "electronvolt", "electronvolts";
         @erg: 1.0_E-7; "erg", "erg", "ergs";
         @foot_poundal: 4.214_011_E-2; "ft · pdl", "foot poundal", "foot poundals";
-        @foot_pound: 1.355_818_E0; "ft · lbf", "foot pound-force", "foot pounds-force"; // @foot_pound_force
+        @foot_pound: 1.355_818_E0; "ft · lbf", "foot pound-force",
+            "foot pounds-force"; // @foot_pound_force
         @kilocalorie_it: 4.186_8_E3; "kcal (IT)", "kilocalorie (IT)", "kilocalories (IT)";
         @kilocalorie: 4.184_E3; "kcal", "kilocalorie", "kilocalories";
-        @kilojoule_hour: 3.6_E6; "kW · h", "kilojoule hour", "kilojoule hours";
+        @kilowatt_hour: 3.6_E6; "kW · h", "kilowatt hour", "kilowatt hours";
         @quad: 1.055_056_E18; "10¹⁵ Btu (IT)", "quad", "quads";
         @therm_ec: 1.055_06_E8; "thm (EC)", "therm (EC)", "therms (EC)";
         @therm_us: 1.054_804_E8; "thm", "therm", "therms";
         @ton_tnt: 4.184_E9; "t of TNT", "ton of TNT", "tons of TNT";
-        @joule_hour: 3.6_E3; "W · h", "joule hour", "joule hours";
-        @joule_second: 1.0_E0; "W · s", "joule second", "joule seconds";
+        @watt_hour: 3.6_E3; "W · h", "watt hour", "watt hours";
+        @watt_second: 1.0_E0; "W · s", "watt second", "watt seconds";
     }
 }
 
@@ -68,7 +73,7 @@ mod tests {
     storage_types! {
         use num::One;
         use si::quantities::*;
-        use si::length as l;
+        use si::area as a;
         use si::mass as m;
         use si::time as t;
         use si::energy as e;
@@ -76,8 +81,9 @@ mod tests {
 
         #[test]
         fn check_dimension() {
-            let _: Energy<V> = Length::new::<l::meter>(V::one()) * Length::new::<l::meter>(V::one()) * Mass::new::<m::kilogram>(V::one())
-                                / (Time::new::<t::second>(V::one()) * Time::new::<t::second>(V::one()));
+            let _: Energy<V> = Area::new::<a::square_meter>(V::one())
+                * Mass::new::<m::kilogram>(V::one())
+                / Time::new::<t::second>(V::one()).powi(::typenum::P2::new());
         }
 
         #[test]
@@ -102,8 +108,9 @@ mod tests {
 
             fn test<M: m::Conversion<V>, E: e::Conversion<V>>() {
                 Test::assert_approx_eq(&Energy::new::<E>(V::one()),
-                    &(Length::new::<l::meter>(V::one()) * Length::new::<l::meter>(V::one()) * Mass::new::<M>(V::one())
-                        / (Time::new::<t::second>(V::one()) * Time::new::<t::second>(V::one()))));
+                    &(Area::new::<a::square_meter>(V::one())
+                        * Mass::new::<M>(V::one())
+                        / Time::new::<t::second>(V::one()).powi(::typenum::P2::new())));
             }
         }
     }
```

Aehmlo commented:
I had been steering away from using derived units in my tests, but I guess as long as those other tests pass, we're good. I've added a lot of tests and units, and fixed some of the common style errors.

iliekturtles commented:
- Electric potential
  - `millivolt` descriptions should be `"millivolt", "millivolts"`.
 - Power
    - Thoughts about simplifying horsepower?
      - Change `horsepower_550` to just `@horsepower: ...; "hp", "horsepower", "horsepower";`. I believe this is the referenced unit when someone just mentions horsepower without qualifiers.
      - Change `horsepower_boiler` abbreviation to `"hp (S)"`. https://en.wikipedia.org/wiki/Horsepower#Definitions
      - Change `horsepower_electric` abbreviation to `"hp (E)"`. https://en.wikipedia.org/wiki/Horsepower#Definitions
      - Change `horsepower` to `@horsepower_metric: ...; "hp (M)", "metric horsepower", "metric horsepower";`
      - Change `horsepower_imperial` to capitalize `Imperial`. Convention used for volume.
      - Change `horsepower_water` to `hydraulic_horsepower`. Not really sure about this one.

iliekturtles commented:
I just pushed some changes to `master` to fix tests on 1.20.0 and reduce build times. Could you rebase this PR on the new master and squash the two commits that affect `electric_current.rs` into one? I'm in the process of reviewing the latest changes in this PR and if I don't find anything there and the build passes once rebased on `master` I'll plan on merging this PR!

iliekturtles commented:
- Electric current
  - I suggested combining the two electric current commits in my previous message. Feel free to leave these separate. I reviewed the changes again and realized two commits is fine.
- Energy
  - `@kilojoule_hour` should be `@kilowatt_hour`.
- Available energy.
  - See diff below for some formatting.
  - Change `check_units` to use three generic parameters: energy, mass, and available energy. Add tests for all available energy units where the energy and mass units exist. `test::<e::joule, m::kilogram, ae::joule_per_kilogram>();`
- Power
  - Wrap `hydraulic_horsepower` at 100 characters.
  - Change `check_units` to use three generic parameters: energy, time, and power. Add tests for all the power units where the energy and time units exist. `test::<e::joule, t::second, p::watt>();`

```diff
diff --git a/src/si/available_energy.rs b/src/si/available_energy.rs
index 9916f05..46b1b0f 100644
--- a/src/si/available_energy.rs
+++ b/src/si/available_energy.rs
@@ -13,28 +13,49 @@ quantity! {
         Z0,     // amount of substance
         Z0>;    // luminous intensity
     units {
-        @yottajoule_per_kilogram: prefix!(yotta); "YJ/kg", "yottajoule per kilogram", "yottajoules per kilogram";
-        @zettajoule_per_kilogram: prefix!(zetta); "ZJ/kg", "zettajoule per kilogram", "zettajoules per kilogram";
-        @exajoule_per_kilogram: prefix!(exa); "EJ/kg", "exajoule per kilogram", "exajoules per kilogram";
-        @petajoule_per_kilogram: prefix!(peta); "PJ/kg", "petajoule per kilogram", "petajoules per kilogram";
-        @terajoule_per_kilogram: prefix!(tera); "TJ/kg", "terajoule per kilogram", "terajoules per kilogram";
-        @gigajoule_per_kilogram: prefix!(giga); "GJ/kg", "gigajoule per kilogram", "gigajoules per kilogram";
-        @megajoule_per_kilogram: prefix!(mega); "MJ/kg", "megajoule per kilogram", "megajoules per kilogram";
-        @kilojoule_per_kilogram: prefix!(kilo); "kJ/kg", "kilojoule per kilogram", "kilojoules per kilogram";
-        @hectojoule_per_kilogram: prefix!(hecto); "hJ/kg", "hectojoule per kilogram", "hectojoules per kilogram";
-        @decajoule_per_kilogram: prefix!(deca); "daJ/kg", "decajoule per kilogram", "decajoules per kilogram";
+        @yottajoule_per_kilogram: prefix!(yotta); "YJ/kg", "yottajoule per kilogram",
+            "yottajoules per kilogram";
+        @zettajoule_per_kilogram: prefix!(zetta); "ZJ/kg", "zettajoule per kilogram",
+            "zettajoules per kilogram";
+        @exajoule_per_kilogram: prefix!(exa); "EJ/kg", "exajoule per kilogram",
+            "exajoules per kilogram";
+        @petajoule_per_kilogram: prefix!(peta); "PJ/kg", "petajoule per kilogram",
+            "petajoules per kilogram";
+        @terajoule_per_kilogram: prefix!(tera); "TJ/kg", "terajoule per kilogram",
+            "terajoules per kilogram";
+        @gigajoule_per_kilogram: prefix!(giga); "GJ/kg", "gigajoule per kilogram",
+            "gigajoules per kilogram";
+        @megajoule_per_kilogram: prefix!(mega); "MJ/kg", "megajoule per kilogram",
+            "megajoules per kilogram";
+        @kilojoule_per_kilogram: prefix!(kilo); "kJ/kg", "kilojoule per kilogram",
+            "kilojoules per kilogram";
+        @hectojoule_per_kilogram: prefix!(hecto); "hJ/kg", "hectojoule per kilogram",
+            "hectojoules per kilogram";
+        @decajoule_per_kilogram: prefix!(deca); "daJ/kg", "decajoule per kilogram",
+            "decajoules per kilogram";
         /// Derived unit of available energy.
-        @joule_per_kilogram: prefix!(none); "J/kg", "joule per kilogram", "joules per kilogram";
-        @decijoule_per_kilogram: prefix!(deci); "dJ/kg", "decijoule per kilogram", "decijoules per kilogram";
-        @centijoule_per_kilogram: prefix!(centi); "cJ/kg", "centijoule per kilogram", "centijoules per kilogram";
-        @millijoule_per_kilogram: prefix!(milli); "mJ/kg", "millijoule per kilogram", "millijoules per kilogram";
-        @microjoule_per_kilogram: prefix!(micro); "┬╡J/kg", "microjoule per kilogram", "microjoules per kilogram";
-        @nanojoule_per_kilogram: prefix!(nano); "nJ/kg", "nanojoule per kilogram", "nanojoules per kilogram";
-        @picojoule_per_kilogram: prefix!(pico); "pJ/kg", "picojoule per kilogram", "picojoules per kilogram";
-        @femtojoule_per_kilogram: prefix!(femto); "fJ/kg", "femtojoule per kilogram", "femtojoules per kilogram";
-        @attojoule_per_kilogram: prefix!(atto); "aJ/kg", "attojoule per kilogram", "attojoules per kilogram";
-        @zeptojoule_per_kilogram: prefix!(zepto); "zJ/kg", "zeptojoule per kilogram", "zeptojoules per kilogram";
-        @yoctojoule_per_kilogram: prefix!(yocto); "yJ/kg", "yoctojoule per kilogram", "yoctojoules per kilogram";
+        @joule_per_kilogram: prefix!(none); "J/kg", "joule per kilogram",
+            "joules per kilogram";
+        @decijoule_per_kilogram: prefix!(deci); "dJ/kg", "decijoule per kilogram",
+            "decijoules per kilogram";
+        @centijoule_per_kilogram: prefix!(centi); "cJ/kg", "centijoule per kilogram",
+            "centijoules per kilogram";
+        @millijoule_per_kilogram: prefix!(milli); "mJ/kg", "millijoule per kilogram",
+            "millijoules per kilogram";
+        @microjoule_per_kilogram: prefix!(micro); "┬╡J/kg", "microjoule per kilogram",
+            "microjoules per kilogram";
+        @nanojoule_per_kilogram: prefix!(nano); "nJ/kg", "nanojoule per kilogram",
+            "nanojoules per kilogram";
+        @picojoule_per_kilogram: prefix!(pico); "pJ/kg", "picojoule per kilogram",
+            "picojoules per kilogram";
+        @femtojoule_per_kilogram: prefix!(femto); "fJ/kg", "femtojoule per kilogram",
+            "femtojoules per kilogram";
+        @attojoule_per_kilogram: prefix!(atto); "aJ/kg", "attojoule per kilogram",
+            "attojoules per kilogram";
+        @zeptojoule_per_kilogram: prefix!(zepto); "zJ/kg", "zeptojoule per kilogram",
+            "zeptojoules per kilogram";
+        @yoctojoule_per_kilogram: prefix!(yocto); "yJ/kg", "yoctojoule per kilogram",
+            "yoctojoules per kilogram";
 
         @joule_per_zeptogram: prefix!(yotta); "J/zg", "joule per zeptogram", "joules per zeptogram";
         @joule_per_attogram: prefix!(zetta); "J/ag", "joule per attogram", "joules per attogram";
@@ -52,9 +73,12 @@ quantity! {
         @joule_per_zettagram: prefix!(atto); "J/Zg", "joule per zettagram", "joules per zettagram";
         @joule_per_yottagram: prefix!(zepto); "J/Yg", "joule per yottagram", "joules per yottagram";
 
-        @btu_it_per_pound: 2.326_E3; "Btu (IT)/lb", "British thermal unit (IT) per pound", "British thermal units (IT) per pound";
-        @btu_per_pound: 2.324_444_E3; "Btu (IT)/lb", "British thermal unit per pound", "British thermal units per pound";
-        @calorie_it_per_pound: 4.186_8_E3; "cal (IT)/lb", "calorie (IT) per pound", "calories (IT) per pound";
+        @btu_it_per_pound: 2.326_E3; "Btu (IT)/lb", "British thermal unit (IT) per pound",
+            "British thermal units (IT) per pound";
+        @btu_per_pound: 2.324_444_E3; "Btu (IT)/lb", "British thermal unit per pound",
+            "British thermal units per pound";
+        @calorie_it_per_pound: 4.186_8_E3; "cal (IT)/lb", "calorie (IT) per pound",
+            "calories (IT) per pound";
         @calorie_per_pound: 4.184_E3; "cal/lb", "calorie per pound", "calories per pound";
     }
 }
@@ -77,7 +101,6 @@ mod tests {
 
         #[test]
         fn check_units() {
-
             test::<a::square_terameter, e::yottajoule_per_kilogram>();
             test::<a::square_gigameter, e::exajoule_per_kilogram>();
             test::<a::square_megameter, e::terajoule_per_kilogram>();
```

Aehmlo commented:
Done (and bumping because I assume you still don't get notified of new commits).
